### ISSUE TITLE
Expose the tree hook

### ISSUE THREAD

**Issue by DeepeshMitra** [ant-design/ant-design#47919]
I have gone through antd features and the discussions and couldnt find something related to this. 
### What problem does this feature solve?

**Context:**
Recently, I wanted to created a nested checklist structure. I wanted to do a lot of customization, e.g.:
1. On the basis of whether a node is checked, unchecked or indeterminate, I wanted to change the color of the node title. 
2. Each of the leaf nodes had a lot of metadata attached to it, which was getting passed into a modal on clicking the leaf node. There was no way for me to easily maintain this metaData other than flattening the entire map myself.

But, in all these cases the basic functionality of checking, unchecking remained the same. For all of these customization, I had to write a separate Provider and wrap the entire component to make use of them. (Which is still fine)

But, the main problem arised when 
1. I wanted to activate a submit button based on whether all the nodes were checked. So, I found no easy way rather than than verifying if the checkedKeys.length == allNodes.length which was a hassle in itself as I had to iterate through the entire tree myself for which I had to maintain a separate tree variable. It would have been great if I could have checked the state of the top nodes only instead of writing a recursive function myself.
2. I had to place the checkboxes somewhere else instead of the tree structure. On discarding the tree sturcuture I lost the benefits of checking, unchecking and updating the state of the tree. If I would have had access to a hook to check a node, it would have been easier.
3. Similarly, I had to place the switcher icon at end of the line instead of the start. On clicking the switcher icon, I would have like to call a function to update the expanded keys. Antd might be internally using it. So, it would have been great if I could use that instead of writing my own.

**Solution** - Expose the nested checklist provider or the hook for it.

**End user experience** - **A developer can customize the nested checklist tree instead of being limited to the tree based structure.**

### What does the proposed API look like?

It would look similar to Form.useForm() and expose apis like 
getNode(nodeKey) => {
isChecked: boolean,
isDisable: boolean,
....
}

setNodeState(key, state) => {
node["key"] = state
}

getNodesOfLevel(level: number) => {
//return all nodes of level 1
}

getExpandedKeys() => {
//Gets all expanded keys
}

setExpandedKeys([expandedKeys]) => {
//sets expanded keys
}

getSelectedKeys() => {
//Gets all selected keys
}

setSelectedKeys([expandedKeys]) => {
//sets selected keys
}

P.S. - Along with keys, the above operations can also be done using pos of the node.

**Comment by afc163** [comment]
Try use Tree in React's way by control `selectedKeys` and `expandedKeys`.

**Comment by DeepeshMitra** [comment]
Thanks for the prompt response. 
Yes, currently I have implemented your suggestion. 
The scope of work was beyond selectedKeys and expandedKeys.
I had implemented provider and other stuff to get the value of selected keys out of the component which was still fine. 

But, once I moved out of the structure of nested tree(still wanted nested checklist, just not the tree) , I was implementing all the functions for checking, unchecking, selecting etc. which I am pretty sure that antd team must have already done internally. Also, I had to implement function to manage the current state of all the nodes.
